Definition
================

    >>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul


Raises
================

::

    >>> print(lazy_matrix_mul("matrix a", [[1, 2]]))
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead

::

    >>> print(lazy_matrix_mul([[1, 2]], "matrix b"))
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead

::

    >>> print(lazy_matrix_mul([[1, 2]], ["matrix b"]))
    Traceback (most recent call last):
    ValueError: shapes (1,2) and (1,) not aligned: 2 (dim 1) != 1 (dim 0)

::

    >>> print(lazy_matrix_mul([], [[2, 2], [1, 1]]))
    Traceback (most recent call last):
    ValueError: shapes (0,) and (2,2) not aligned: 0 (dim 0) != 2 (dim 0)

::

    >>> print(lazy_matrix_mul([[]], [[2, 2], [1, 1]]))
    Traceback (most recent call last):
    ValueError: shapes (1,0) and (2,2) not aligned: 0 (dim 1) != 2 (dim 0)

::

    >>> print(lazy_matrix_mul([[2, 2], [1, 1]], []))
    Traceback (most recent call last):
    ValueError: shapes (2,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

::

    >>> print(lazy_matrix_mul([[2, 2], [1, 1]], [[]]))
    Traceback (most recent call last):
    ValueError: shapes (2,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

::

    >>> print(lazy_matrix_mul([["#", 2], [1, 1]], [[1],[2],[3]]))
    Traceback (most recent call last):
    ValueError: shapes (2,2) and (3,1) not aligned: 2 (dim 1) != 3 (dim 0)

::

    >>> print(lazy_matrix_mul([[3, 2], [1, 1]], [[1, "t", 2], [2, 3, 4]]))
    Traceback (most recent call last):
    TypeError: invalid data type for einsum

::

    >>> print(lazy_matrix_mul([[3, 2], [1]], [[1, 1, 2], [2, 3, 4]]))
    Traceback (most recent call last):
    ValueError: setting an array element with a sequence.

::

    >>> print(lazy_matrix_mul([[3, 2], [2, 1]], [[1, 1, 2], [3, 4]]))
    Traceback (most recent call last):
    ValueError: setting an array element with a sequence.

::

    >>> print(lazy_matrix_mul([[2, 4], [2, 2], [1, 2]], [[4, 4], [2, 2], [1, 1]]))
    Traceback (most recent call last):
    ValueError: shapes (3,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)

Cases
================

::

    >>> print(lazy_matrix_mul([[1, 2]], [[1], [5]]))
    [[11]]
    >>> print(lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]]))
    [[ 7 10]
     [15 22]]
    >>> print(lazy_matrix_mul([[1, 2]], [[3, 4], [5, 6]]))
    [[13 16]]
    >>> print(lazy_matrix_mul([[1, 2], [3, 4]], [[5, 6], [7, 8]]))
    [[19 22]
     [43 50]]
